platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/StoreExecutionStateStep.java

```diff
@@ -47,7 +49,7 @@ public R execute(UnitOfWork work, C context) {
                 .flatMap(beforeExecutionState -> result.getAfterExecutionOutputState()
                     .filter(afterExecutionState -> result.getExecution().isSuccessful() || shouldPreserveFailedState(context, afterExecutionState))
                     .map(executionOutputState -> new DefaultAfterExecutionState(beforeExecutionState, executionOutputState)))
-                .ifPresent(afterExecutionState -> history.store(context.getIdentity().getUniqueId(), afterExecutionState)));
+                .ifPresent(afterExecutionState -> history.store(context.getIdentity().getUniqueId(), Objects.requireNonNull(context.getCacheKey()), afterExecutionState)));
```

Comment:
🤔 This is the part I'm unhappy about, but let's encode the "no cache key available" case in the context type hierarchy at a later time. Could you please add a TODO here?

Reply:
I added a TODO.